modules/snmp-dest/snmpdest.c

```diff
@@ -0,0 +1,899 @@
+  self->host_port = g_string_sized_new(64);
+  if (self->transport)
+    g_string_append_printf(self->host_port, "%s:", self->transport);
+  g_string_append_printf(self->host_port, "%s:%d", self->host, self->port);
+
+  self->queue = log_dest_driver_acquire_queue(&self->super, snmpdest_dd_format_persist_name(self));
+
+  stats_lock();
+  stats_cluster_logpipe_key_set(&sc_key_queued, SCS_SNMP, snmpdest_dd_format_stats_instance(self), NULL );
+  stats_register_counter(1, &sc_key_queued, SC_TYPE_QUEUED, &self->queued_messages);
+
+  stats_cluster_logpipe_key_set(&sc_key_dropped, SCS_SNMP, snmpdest_dd_format_stats_instance(self), NULL );
+  stats_register_counter(1, &sc_key_dropped, SC_TYPE_DROPPED, &self->dropped_messages);
+
+  stats_cluster_logpipe_key_set(&sc_key_processed, SCS_SNMP, snmpdest_dd_format_stats_instance(self), NULL );
+  stats_register_counter(1, &sc_key_processed, SC_TYPE_PROCESSED, &self->processed_messages);
+  stats_unlock();
+
+  msg_verbose("Initializing SNMP destination",
+              evt_tag_str("host", self->host),
+              evt_tag_int("port", self->port),
+              NULL);
+
+  GlobalConfig *cfg = log_pipe_get_config(s);
+  gchar **tz = &self->template_options.time_zone[LTZ_LOCAL],
+          *time_zone = cfg->template_options.time_zone[LTZ_SEND];
+  if (!*tz && time_zone)
+    /* inherit the global set_time_zone if the local one hasn't been set */
+    *tz = g_strdup(time_zone);
+  log_template_options_init(&self->template_options, cfg);
+  /* set the cfg field of the log template items */
+  GList *t = self->snmp_templates;
+  while (t)
+    {
+      ((LogTemplate *)t->data)->cfg = cfg;
+      t = t->next;
+    }
+
+  snmpdest_dd_session_init(self);
+
+  snmpdest_dd_start_thread(self);
+  return TRUE;
+}
+
+static gboolean
+snmpdest_dd_deinit(LogPipe *s)
+{
+  SNMPDestDriver *self = (SNMPDestDriver *)s;
+
+  snmpdest_dd_stop_thread(self);
+
+  /* SNMP deinit */
+  if (self->session_initialized)
+    {
+      snmp_close(self->ss);
+      free_session(&self->session);
+    }
+
+  stats_lock();
+  stats_unregister_counter(&sc_key_queued, SC_TYPE_QUEUED, &self->queued_messages);
+  stats_unregister_counter(&sc_key_processed, SC_TYPE_PROCESSED, &self->processed_messages);
+  stats_unregister_counter(&sc_key_dropped, SC_TYPE_DROPPED, &self->dropped_messages);
+  stats_unlock();
+
+  if (!log_dest_driver_deinit_method(s))
+    return FALSE;
+
+  return TRUE;
+}
+
+static gint snmp_dest_counter = 0;
+
+static void
+snmpdest_dd_free(LogPipe *d)
+{
+  SNMPDestDriver *self = (SNMPDestDriver *)d;
+
+  if (snmp_dest_counter == 1)
+    {
+      /* this is the last snmp driver, close the snmp */
+      SOCK_CLEANUP;
+      snmp_shutdown(s_snmp_name);
+    }
+  --snmp_dest_counter;
+
+  g_mutex_free(self->queue_mutex);
+  g_cond_free(self->writer_thread_wakeup_cond);
+
+  g_free(self->version);
+  g_free(self->host);
+  if (self->host_port)
+    g_string_free(self->host_port, TRUE);
+  while(self->snmp_objs)
```

Comment:
g_list_free_full might be simpler.

Reply:
g_list_free_full added to all three lists